### ISSUE TITLE
[no-floating-promises] Ability to ignore certain types from producing errors

### ISSUE THREAD

**Issue by blimmer** [typescript-eslint/typescript-eslint#2640]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

To support this issue, I've created [a sample repository](https://github.com/blimmer/eslint-no-floating-promises) to show the problem. The main issue is with types that support a "builder pattern". In other words, you can continually build onto a given object until you're ready to execute it with a `.then` or `await`.

```JSON
{
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

There are two issues displayed in this use-case.

1. **Fastify**. Fastify causes two classes of errors to be thrown in this example. The first is with `.register`. This one can pretty easily be ignored by adding a `void` in front of it. However, all of the Fastify examples don't show the need to `await` the returned promises. The second issue is with their `FastifyReply` object, which is likely more interesting to think about. It uses a builder pattern, allowing you to set properties before your file `res.send`. Each line adds a warning before the final `send`, unless you add `void`s.

2. **Knex QueryBuilder**. The other issue displayed here is with the Knex QueryBuilder. Knex allows you to pass around a `QueryBuilder` object before finally `await`-ing it to execute the query. Again, these errors can be supressed with a `void` until you're actually ready to execute the query.

```TS
import Fastify from 'fastify';
import FastifyCookie from 'fastify-cookie';
import knex from 'knex';

const dbClient = knex({});
const app = Fastify();

// The docs for Fastify don't show a need to `await` these calls to `register`.
// It would be nice to be able to safelist these usages somehow.
app.register(FastifyCookie);

app.get('/example', (_req, res) => {
  // You're able to build up your response over multiple lines. These throw
  // linting errors since the FastifyReply type has a `then` method.
  res.header("x-some-header", "foobar");
  res.send(200);
})

app.get('/dbquery', async (_req, res) => {
  // This one is a bit tougher to think about solving for, since eventually
  // you'd want to validate that it's `await`-ed. However, it would still be
  // nice to provide some way to just ignore these types all together.
  const query = dbClient('foos');
  query.orderBy('id', 'desc');
  const dbResult = await query;
  res.send(JSON.stringify(dbResult));
});

app.listen(4000);
```

```json
{
  "compilerOptions": {
    "target": "es2019",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true
  }
}
```

**Expected Result**

It would be nice to be able to ignore certain types from producing errors. In this case, it'd be nice to be able to safelist the `FastifyReply` type from producing warnings in this plugin.

**Actual Result**

```console
  10:1  error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
  15:3  error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
  16:3  error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
  24:3  error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
  26:3  error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
  29:1  error  Promises must be handled appropriately or explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
```

**Additional Info**

I can work around this by placing `void` in front of each of the calls I want to ignore. However, this is more of a feature request/brainstorming issue to think about ways we could reduce cluttering the code with `void` statements in all of these cases.

Is this even something to consider fixing? Or maybe adding the `void` in front of all calls to builder methods is explicit and desired?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.0` |
| `@typescript-eslint/parser`        | `4.4.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.10.0` |
| `node`                             | `12.18.3` |

**Comment by bradzacher** [comment]
Let me just begin by saying in no uncertain terms - no to your request.

Ignore lists are bad solutions to linting problems because they lack any context - which means they have very real chances to be misused and cause false _negatives_.  
False negatives are worse for the ecosystem than false positives, because missing a problem can leak bugs into your code, and that erodes the user's trust in the linter.

---

> However, all of the Fastify examples don't show the need to await the returned promises.

Technically there's never any need to await a returned promise, unless you want to use its result or do things synchronously.

Just because the examples don't `await` them doesn't make them any less of floating promises!

---

The reason that `FastifyReply` is erroring here is due to the fact that it declares a `then` method.
https://github.com/fastify/fastify/blob/341bbc86c94df5cec49778e9c5657b27cbf59282/types/reply.d.ts#L48

`QueryBuilder` has the same problem, except here it just _copies_ the method definitions from `Promise`:
https://github.com/knex/knex/blob/80355e48509857fb4941f65f4b611973a9ee99d5/types/index.d.ts#L1475-L1483

The rule currently just detects if the type in question has a `then` which looks the same as a promise's `then`.

https://github.com/typescript-eslint/typescript-eslint/blob/6bc93257ec876214743a165093b6666d713379f6/packages/eslint-plugin/src/rules/no-floating-promises.ts#L175-L197

I don't know the best solution here, as there's nothing that solves all of the problems in a safe way.

If the rule only considered types that extend `Promise`, that would solve the Fastify problem.
However then it would never match on `QueryBuilder` - which is bad because you should be awaiting that at some point.
Also this would never match libraries like `Bluebird`.

**Comment by blimmer** [comment]
To start off, fair enough on not wanting to implement this behavior. As I mentioned in my original comment, this was mostly opened for brainstorming and consideration. I appreciate your time in reviewing this PR and for your descriptive comment. Tone is tough to decipher in text, but I guess I want to head off any (perceived) negativity in this thread as we continue on.

Perhaps in these cases, it'd be more appropriate to open discussion about these patterns in the respective projects. For instance, I wonder if `knex` would consider having an explicit `execute()` method to use instead of inferring execution when the promise is `.then()`-ed or `await`ed.
